src/sentry/utils/appleconnect/itunes_connect.py

```diff
@@ -3,349 +3,515 @@
+
+    def sms_code(self, code: str) -> None:
+        """Sends the SMS auth code, completing authentication.
+
+        :raises InvalidSmsAuthError:
+        """
+        assert self.state is ClientState.SMS_AUTH_REQUESTED
+        assert self._trusted_phone is not None
+        url = "https://idmsa.apple.com/appleauth/auth/verify/phone/securitycode"
+        logger.debug("PUT {url}")
+        response = self.session.post(
+            url,
+            json={
+                "securityCode": {"code": code},
+                "phoneNumber": {"id": self._trusted_phone.id},
+                "mode": self._trusted_phone.push_mode,
+            },
+            headers={
+                "scnt": self._scnt,
+                "X-Apple-Id-Session-Id": self.session_id,
+                "Accept": "application/json",
+                "X-Apple-Widget-Key": self.service_key,
+            },
+        )
+        if response.status_code != HTTPStatus.OK:
+            # TODO: Make invalid code distinguishable from generic error.
+            raise InvalidAuthCodeError
+        else:
+            self.state = ClientState.AUTHENTICATED
+
+        # If we want more info about the failure extract failure info from the response body
+        # response is a JSON with the following interesting fields:
+        # (Extracted from a login attempt with wrong code)
+        # 'restrictedAccount': False,
+        # 'securityCode': {'code': '123123',
+        #                  'securityCodeCooldown': False,
+        #                  'securityCodeLocked': False,
+        #                  'tooManyCodesSent': False,
+        #                  'tooManyCodesValidated': False},
+        # 'serviceErrors': [{'code': '-21669',
+        #                    'message': 'Incorrect verification code.',
+        #                    'suppressDismissal': False,
+        #                    'title': 'Incorrect Verification Code'}],
+
+    def session_cookie(self) -> str:
+        """Extracts the session cookies.
+
+        This can be used to create a new client without having to re-authenticate, if the
+        session is still valid.
+
+        :returns: the session cookie.
+        :raises: an exception if the client is not yet authenticated and has no session
+           cookie.
+        """
+        assert self.state is ClientState.AUTHENTICATED
+        return self.session.cookies.get(SESSION_COOKIE_NAME)  # type: ignore
+
+    def load_session_cookie(self, cookie: str) -> None:
+        """Loads the itunes session cookie in the client's session.
+
+        This will invalidate the client's authentication state and reset it.  It will
+        trigger a request to check a cookie is valid and use this to set the state
+        correctly.
+
+        If the iTunes session is still valid the user will be logged in inside the session.
+        """
+        self.session.cookies.set(self.SESSION_COOKIE_NAME, cookie)  # type: ignore
+        try:
+            self._request_session_info()
```

Comment:
what does it mean to reset a client's authentication state? should `load_session_cookie` (or `_request_session_info`) be resetting the properties on `self` that are set during the authentication process such as `session_id`, `_scnt`, and `_trusted_phone`? 

is this line only interested in the side effects of `self._request_session_info()`?

Reply:
none of the attributes really affect the state at this point.  The new state will be either AUTHENTICATED or EXPIRED and neither of those states care about any of those attributes.  The only way to start reading those attributes again is by going through the states that set them first.